---
title: "Class06: R Functions"
author: "Nicole Lin (PID: A17826971)"
format: pdf
toc: true
---

## Background

Functions are at the heart of using R. Everything we do involves calling and using functions (from data input, analysis, to results output).

All functions in R have at least 3 things:

1. A **name** the thing we to call the function.
2. One or more input **arguments** that are comma-separated.
3. The **body**, lines of code between curly brackets { } that does the work of the function.

## A first function

Let's write a silly wee function to add some numbers:

```{r}
add <- function(x) {
  x + 1
}
```

Let's try it out
```{r}
add(100)
```

Will this work?

```{r}
add(c(100, 200, 300))
```
Modify to be more useful and add more than just 1.

```{r}
add <- function(x, y=1) {
  x + y
}
```

```{r}
add(100, 10)
```

Will this work?
```{r}
add(100)
```

```{r}
plot(1:10, col="blue", typ="b")
```

```{r}
log(10, base=10)
```

> **N.B.** Input arguments can be either **required** or **optional**. The latter has a fall-back default that is specified in the function code with an equal sign.


```{r}
#add(100, 200, 300)
```


## A second function

All functions in R look like this

```
name <- function(arg) {
  body
}
```

The `sample()` function in R randomly selects a specified amount of items from a vector with or without replacement. With 4 arguments: x, size, replace, and prob (x and size are required). Default for replace is FALSE (without replacement)

```{r}
sample(1:10, size=4)
```

> Q. Return 12 numbers picked randomly from the input 1:10

```{r}
sample(1:10, size=12, replace=TRUE)
```

> Q. Write the code to generate a random 12 nucleotide long DNA sequence


```{r}
sample(x= c("A", "T", "G", "C"), size=12, replace = TRUE)
```


> Q. Write a first-version function called `generate_dna()` that generates a user specified length `n` random DNA sequence?

```
name <- function(arg) {
  body
}
```

``` {r}
generate_dna <- function(n=6) {
  bases <- c("A", "T", "G", "C")
  sample(bases, size=n, replace=TRUE)
}
```


```{r}
generate_dna(100)
```

> Q. Modify your function to return a FASTA-like sequence so rather than [1] "G" "C" "A" "A" "T" we want "GCAAT"

``` {r}
generate_dna <- function(n=6) {
  bases <- c("A", "T", "G", "C")
  ans <- sample(bases, size=n, replace=TRUE)
  paste(ans, collapse = "")
}
```


```{r}
generate_dna(10)
```

> Q. Give the user an option to return FASTA format output sequence or standard multi-element vector format.

``` {r}
generate_dna <- function(n=6, fasta=TRUE) {
  bases <- c("A", "T", "G", "C")
  ans <- sample(bases, size=n, replace=TRUE)
 
  if(fasta) {
    ans <- paste(ans, collapse = "")
    cat("Hello...")
  } else {
    cat("...is it me you are looking for...")
  }
  
  return(ans)
}
```

```{r}
generate_dna(10)
generate_dna(10, fasta=F)
```

## A new cool function

> Q. Write a function called `generate_protein()` that generates a user specified length protein sequence in FASTA-like format.



```{r}
generate_protein <- function(n=6) {
  aa <- c("A","R","N","D","C",
        "Q","E","G","H","I",
        "L","K","M","F","P",
        "S","T","W","Y","V")
  ans <- sample(aa, size=n, replace=TRUE)
  ans <- paste(ans, collapse = "")
  return(ans)
}
```

```{r}
generate_protein(30)
```


> Q. Use your new `generate_protein()` function to generate all sequences between length 6 and 12 amino-acids in length and check if any of these are unique in nature (i.e. found in the NR database at NCBI).

We could de a `for()` loop:

```{r}
for(i in 6:12) {
  cat(">", i, sep="", "\n")
  cat( generate_protein(i), "\n" )
}
```





